File: Aeds2/TP02/Games/AVLGames.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_LINE 5000
#define MAX_TOKEN 500
#define LOG_FILE "845860_arvoreAVL.txt"

typedef struct {
    char App_Id[50];
    char Name[300];
    char Release_Date[50];
    char Estimated_Owners[50];
    double Price;
    char Supported_Languages[500];
    int Metacritic_Score;
    double User_Score;
    int Achievements;
    char Publishers[200];
    char Developers[200];
    char Categories[500];
    char Genres[500];
    char Tags[500];
} AVLGames;

typedef struct Node {
    AVLGames game;
    struct Node *left;
    struct Node *right;
    int height;
} Node;

typedef struct {
    Node *root;
    int comparacoes;
} Tree;

int max(int a, int b) { return (a > b ? a : b); }

int height(Node *n) {
    return n ? n->height : 0;
}

int getBalance(Node *n) {
    return n ? height(n->left) - height(n->right) : 0;
}

Node* newNode(AVLGames game) {
    Node *n = (Node*)malloc(sizeof(Node));
    n->game = game;
    n->left = n->right = NULL;
    n->height = 1;
    return n;
}

Node* rightRotate(Node *y) {
    Node *x = y->left;
    Node *T2 = x->right;

    x->right = y;
    y->left = T2;

    y->height = max(height(y->left), height(y->right)) + 1;
    x->height = max(height(x->left), height(x->right)) + 1;

    return x;
}

Node* leftRotate(Node *x) {
    Node *y = x->right;
    Node *T2 = y->left;

    y->left = x;
    x->right = T2;

    x->height = max(height(x->left), height(x->right)) + 1;
    y->height = max(height(y->left), height(y->right)) + 1;

    return y;
}

Node* insertRec(Tree *t, Node *node, AVLGames game) {

    if (node == NULL) return newNode(game);

    t->comparacoes++;
    int cmp = strcmp(game.Name, node->game.Name);

    if (cmp < 0)
        node->left = insertRec(t, node->left, game);

    else if (cmp > 0)
        node->right = insertRec(t, node->right, game);

    else
        return node;

    node->height = 1 + max(height(node->left), height(node->right));

    int balance = getBalance(node);

    if (balance > 1 && strcmp(game.Name, node->left->game.Name) < 0)
        return rightRotate(node);

    if (balance < -1 && strcmp(game.Name, node->right->game.Name) > 0)
        return leftRotate(node);

    if (balance > 1 && strcmp(game.Name, node->left->game.Name) > 0) {
        node->left = leftRotate(node->left);
        return rightRotate(node);
    }

    if (balance < -1 && strcmp(game.Name, node->right->game.Name) < 0) {
        node->right = rightRotate(node->right);
        return leftRotate(node);
    }

    return node;
}

void insert(Tree *t, AVLGames game) {
    t->root = insertRec(t, t->root, game);
}


void searchRec(Tree *t, Node *node, char *name) {
    if (node == NULL) {
        printf("NAO\n");
        return;
    }

    t->comparacoes++;
    int cmp = strcmp(name, node->game.Name);

    if (cmp == 0) {
        printf("SIM\n");
        return;
    }

    if (cmp < 0) {
        printf("esq ");
        searchRec(t, node->left, name);
    } else {
        printf("dir ");
        searchRec(t, node->right, name);
    }
}

void searchPrint(Tree *t, char *name) {
    printf("raiz ");
    searchRec(t, t->root, name);
}


void lerCSV(AVLGames *g, char *linha) {
    char temp[MAX_LINE];
    strcpy(temp, linha);

    char *token;
    int coluna = 0;

    char *ptr = temp;

    token = strtok(ptr, ",");
    while (token != NULL && coluna < 14) {

        switch (coluna) {
            case 0: strcpy(g->App_Id, token); break;
            case 1: strcpy(g->Name, token); break;
            case 2: strcpy(g->Release_Date, token); break;
            case 3: strcpy(g->Estimated_Owners, token); break;
            case 4: g->Price = atof(token); break;
            case 5: strcpy(g->Supported_Languages, token); break;
            case 6: g->Metacritic_Score = atoi(token); break;
            case 7: g->User_Score = atof(token); break;
            case 8: g->Achievements = atoi(token); break;
            case 9: strcpy(g->Publishers, token); break;
            case 10: strcpy(g->Developers, token); break;
            case 11: strcpy(g->Categories, token); break;
            case 12: strcpy(g->Genres, token); break;
            case 13: strcpy(g->Tags, token); break;
        }

        token = strtok(NULL, ",");
        coluna++;
    }
}

int main() {

    clock_t inicio = clock();

    Tree arvore;
    arvore.root = NULL;
    arvore.comparacoes = 0;

    FILE *f = fopen("/tmp/games.csv", "r");
    if (!f) {
        printf("Erro abrindo /tmp/games.csv\n");
        return 1;
    }

    char linha[MAX_LINE];
    fgets(linha, MAX_LINE, f); 

    char **map_ids = NULL;
    char **map_linhas = NULL;
    int map_count = 0;

    while (fgets(linha, MAX_LINE, f)) {

        char id[50];
        int i = 0;
        while (linha[i] != ',' && linha[i] != '\0') {
            id[i] = linha[i];
            i++;
        }
        id[i] = '\0';

        map_count++;
        map_ids = realloc(map_ids, map_count * sizeof(char*));
        map_linhas = realloc(map_linhas, map_count * sizeof(char*));

        map_ids[map_count-1] = strdup(id);
        map_linhas[map_count-1] = strdup(linha);
    }
    fclose(f);

    char entrada[300];

    while (1) {
        fgets(entrada, sizeof(entrada), stdin);
        entrada[strcspn(entrada, "\r\n")] = 0;

        if (strcmp(entrada, "FIM") == 0) break;

        for (int i = 0; i < map_count; i++) {
            if (strcmp(map_ids[i], entrada) == 0) {

                AVLGames g;
                lerCSV(&g, map_linhas[i]);
                insert(&arvore, g);
                break;
            }
        }
    }

    while (1) {
        if (!fgets(entrada, sizeof(entrada), stdin)) break;

        entrada[strcspn(entrada, "\r\n")] = 0;
        if (strcmp(entrada, "FIM") == 0) break;

        printf("%s: ", entrada);
        searchPrint(&arvore, entrada);
    }

    double tempo = (double)(clock() - inicio) / CLOCKS_PER_SEC;

    FILE *log = fopen(LOG_FILE, "w");
    fprintf(log, "845860\t%d\t%.3f", arvore.comparacoes, tempo);
    fclose(log);

    return 0;
}

File: Aeds2/TP02/Games/AVLGames.java
import java.io.*;
import java.text.*;
import java.util.*;

public class AVLGames implements Cloneable {

    private String App_Id;
    private String Name;
    private Date Release_Date;
    private String Estimated_Owners;
    private double Price;
    private String[] Supported_Languages;
    private int Metacritic_Score;
    private double User_Score;
    private int Achievements;
    private String Publishers;
    private String Developers;
    private String[] Categories;
    private String[] Genres;
    private String[] Tags;

    AVLGames() {
        this.App_Id = "NaN";
        this.Name = "NaN";
        this.Release_Date = null;
        this.Estimated_Owners = "NaN";
        this.Price = 0.0;
        this.Supported_Languages = new String[]{"NaN"};
        this.Metacritic_Score = -1;
        this.User_Score = -1.0;
        this.Achievements = 0;
        this.Publishers = "NaN";
        this.Developers = "NaN";
        this.Categories = new String[]{"NaN"};
        this.Genres = new String[]{"NaN"};
        this.Tags = new String[]{"NaN"};
    }

    public void lerCSV(String linha) {
        String[] objetos = new String[14];
        String[] separar = linha.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)", -1);

        for (int i = 0; i < objetos.length; i++) {
            objetos[i] = (i < separar.length && !separar[i].trim().isEmpty())
                    ? separar[i].trim().replaceAll("^\"|\"$", "")
                    : "NaN";
        }

        SimpleDateFormat sdfOriginDateFormat = new SimpleDateFormat("MMM dd, yyyy", Locale.ENGLISH);

        this.App_Id = objetos[0];
        this.Name = objetos[1];

        try {
            this.Release_Date = objetos[2].equals("NaN") ? null : sdfOriginDateFormat.parse(objetos[2]);
        } catch (Exception e) {
            this.Release_Date = null;
        }

        this.Estimated_Owners = objetos[3];
        this.Price = objetos[4].equals("NaN") ? 0.0 : Double.parseDouble(objetos[4]);
        this.Supported_Languages = objetos[5].equals("NaN") ?
                new String[]{"NaN"} : objetos[5].replaceAll("\\[|\\]|'", "").split(",\\s*");

        this.Metacritic_Score = objetos[6].equals("NaN") ? -1 : Integer.parseInt(objetos[6]);
        this.User_Score = objetos[7].equals("NaN") ? -1.0 : Double.parseDouble(objetos[7]);
        this.Achievements = objetos[8].equals("NaN") ? 0 : Integer.parseInt(objetos[8]);
        this.Publishers = objetos[9];
        this.Developers = objetos[10];
        this.Categories = objetos[11].equals("NaN") ? new String[]{"NaN"} : objetos[11].split(",");
        this.Genres = objetos[12].equals("NaN") ? new String[]{"NaN"} : objetos[12].split(",");
        this.Tags = objetos[13].equals("NaN") ? new String[]{"NaN"} : objetos[13].split(",");
    }

    public String getName() {
        return this.Name;
    }

    public Integer getReleaseYear() {
        if (this.Release_Date == null) return null;
        Calendar cal = Calendar.getInstance();
        cal.setTime(this.Release_Date);
        return cal.get(Calendar.YEAR);
    }

    static class Node {
        AVLGames game;
        Node left, right, parent;
        boolean color;

        Node(AVLGames g) {
            this.game = g;
            this.color = true; 
        }
    }
    static class RBT {
        Node root;
        int comparacoes = 0;

        private void rotateLeft(Node x) {
            if (x == null || x.right == null) return;

            Node y = x.right;
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }

            y.parent = x.parent;
            if (x.parent == null) {
                root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }

            y.left = x;
            x.parent = y;
        }

        private void rotateRight(Node x) {
            if (x == null || x.left == null) return;

            Node y = x.left;
            x.left = y.right;
            if (y.right != null) {
                y.right.parent = x;
            }

            y.parent = x.parent;
            if (x.parent == null) {
                root = y;
            } else if (x == x.parent.right) {
                x.parent.right = y;
            } else {
                x.parent.left = y;
            }

            y.right = x;
            x.parent = y;
        }

        void insert(AVLGames g) {
            Node novo = new Node(g);
            if (root == null) {
                root = novo;
                root.color = false; 
                return;
            }

            Node current = root;
            Node parent = null;

            while (current != null) {
                parent = current;
                comparacoes++;
                int cmp = novo.game.getName().compareTo(current.game.getName());
                if (cmp < 0) {
                    current = current.left;
                } else if (cmp > 0) {
                    current = current.right;
                } else {
                    return;
                }
            }

            novo.parent = parent;
            comparacoes++;
            if (novo.game.getName().compareTo(parent.game.getName()) < 0)
                parent.left = novo;
            else
                parent.right = novo;

            fixInsert(novo);
        }

        private void fixInsert(Node z) {
            while (z.parent != null && z.parent.color == true) {
                Node gp = z.parent.parent;
                if (gp == null) break; 

                if (z.parent == gp.left) {
                    Node y = gp.right; 
                    if (y != null && y.color == true) {
                        z.parent.color = false;
                        y.color = false;
                        gp.color = true;
                        z = gp;
                    } else {
                        if (z == z.parent.right) {
                            z = z.parent;
                            rotateLeft(z);
                        }
                        z.parent.color = false;
                        gp.color = true;
                        rotateRight(gp);
                    }
                } else {
                    Node y = gp.left; 
                    if (y != null && y.color == true) {
                        z.parent.color = false;
                        y.color = false;
                        gp.color = true;
                        z = gp;
                    } else {
                        if (z == z.parent.left) {
                            z = z.parent;
                            rotateRight(z);
                        }
                        z.parent.color = false;
                        gp.color = true;
                        rotateLeft(gp);
                    }
                }
            }
            if (root != null) root.color = false; 
        }

        void searchPrint(String name) {
            System.out.print("=>raiz  ");
            searchRec(root, name);
        }

        private void searchRec(Node node, String name) {
            if (node == null) {
                System.out.println("NAO");
                return;
            }

            comparacoes++;
            int cmp = name.compareTo(node.game.getName());

            if (cmp == 0) {
                System.out.println("SIM");
                return;
            }

            if (cmp < 0) {
                System.out.print("esq ");
                searchRec(node.left, name);
            } else {
                System.out.print("dir ");
                searchRec(node.right, name);
            }
        }
    }

    public static void main(String[] args) throws Exception {

        long inicio = System.currentTimeMillis();

        Map<String, String> mapCSV = new HashMap<>();
        BufferedReader br = new BufferedReader(new FileReader("/tmp/games.csv"));
        br.readLine();
        String linha;

        while ((linha = br.readLine()) != null) {
            String id = linha.split(",")[0];
            mapCSV.put(id, linha);
        }
        br.close();

        Scanner sc = new Scanner(System.in);
        RBT arvore = new RBT();

        while (true) {
            String entrada = sc.nextLine();
            if (entrada.equals("FIM")) break;

            if (mapCSV.containsKey(entrada)) {
                AVLGames game = new AVLGames();
                game.lerCSV(mapCSV.get(entrada));
                arvore.insert(game);
            }
        }

        while (sc.hasNextLine()) {
            String nome = sc.nextLine();
            if (nome.equals("FIM")) break;

            System.out.print(nome + ": ");
            arvore.searchPrint(nome);
        }

        sc.close();

        long fim = System.currentTimeMillis();
        double tempo = (fim - inicio) / 1000.0;

        FileWriter log = new FileWriter("845860_arvoreAlvinegra.txt");
        log.write("845860\t" + arvore.comparacoes + "\t" + tempo);
        log.close();
    }
}
